### PR TITLE
RavenDB-20131 add Ipv6 support in Setup

### PR DESCRIPTION
**Pull request by lukaszdobrzynski**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20131/IPv6-support-in-Setup

### Additional description

The update allows setting up a cluster using Ipv4 and Ipv6 addresses. The `api.ravendb.net` service has already been adjusted in this respect. This means users can claim a domain being resolved to both A and AAAA record types.   

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes 
- [x] No

### UI work

- [x] It requires further work in the Studio. See https://issues.hibernatingrhinos.com/issue/RavenDB-23316
- [ ] No UI work is needed

### CODE REVIEW

**Review by gregolsky — APPROVED** [review]

**Review by hiddenshadow21 — APPROVED** [review]

File: src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs [hunk 1]
```diff
@@ -169,62 +169,180 @@ public static async Task UpdateDnsRecordsTask(UpdateDnsRecordParameters paramete
...
+            AssertExpectedIpsEqualLocalIps(allLocalIps);
+        }
 
-                var responseString = await response.Content.ReadAsStringWithZstdSupportAsync(cts.Token).ConfigureAwait(false);
+        void AssertExpectedIpsEqualGoogleIps(HashSet<string> actualGoogleIps)
+        {
+            var allExpectedIps = GetAllExpectedIps();
+            
+            if (actualGoogleIps.SetEquals(allExpectedIps) == false)
+                throw new InvalidOperationException($"Tried to resolve '{hostname}' using Google's api ({GoogleDnsApi}).{Environment.NewLine}" +
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs]: api => API

File: src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs [hunk 2]
```diff
@@ -169,62 +169,180 @@ public static async Task UpdateDnsRecordsTask(UpdateDnsRecordParameters paramete
...
+                                                    $"Expected to get these ips: {string.Join(", ", allExpectedIps)} while Google's actual result was: {string.Join(", ", actualGoogleIps)}" +
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs]: ips => IP's

File: src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs [hunk 3]
```diff
@@ -169,62 +169,180 @@ public static async Task UpdateDnsRecordsTask(UpdateDnsRecordParameters paramete
...
+                                                    "Please wait a while until DNS propagation is finished and try again. If you are trying to update existing DNS records, it might take hours to update because of DNS caching. If the issue persists, contact RavenDB's support.");
+        }
+
+        void AssertExpectedIpsEqualLocalIps(HashSet<string> actualLocalIps)
+        {
+            var allExpectedIps = GetAllExpectedIps();
+            
+            if (allExpectedIps.SetEquals(actualLocalIps) == false)
+                throw new InvalidOperationException($"Tried to resolve '{hostname}' locally but got an outdated result." +
+                                                    Environment.NewLine + $"Expected to get these ips: {string.Join(", ", allExpectedIps)} while the actual result was: {string.Join(", ", actualLocalIps)}" +
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs]: ips => IP's

File: src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs [hunk 4]
```diff
@@ -169,62 +169,180 @@ public static async Task UpdateDnsRecordsTask(UpdateDnsRecordParameters paramete
...
+                                                    Environment.NewLine + $"If we try resolving through Google's api ({GoogleDnsApi}), it works well." +
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs]: api => API

File: src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs [hunk 5]
```diff
@@ -169,62 +169,180 @@ public static async Task UpdateDnsRecordsTask(UpdateDnsRecordParameters paramete
...
+            var expectedIpsV4 = expectedAddresses
+                .Where(x => x.Address.AddressFamily == AddressFamily.InterNetwork)
+                .Select(address => address.Address.ToString())
+                .ToHashSet();
+            var expectedIpsV6 = expectedAddresses
+                .Where(x => x.Address.AddressFamily == AddressFamily.InterNetworkV6)
+                .Select(address => address.Address.ToString())
+                .ToHashSet();
+            
+            return expectedIpsV4.Concat(expectedIpsV6).ToHashSet();
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs]: isnt that

```csharp
            return expectedAddresses
                .Where(x => x.Address.AddressFamily == AddressFamily.InterNetwork || x.Address.AddressFamily == AddressFamily.InterNetworkV6)
                .Select(address => address.Address.ToString())
                .ToHashSet();
```

File: src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs [hunk 6]
```diff
@@ -169,62 +169,180 @@ public static async Task UpdateDnsRecordsTask(UpdateDnsRecordParameters paramete
...
+        }
+
+        async Task<HashSet<string>> GetGoogleDnsResult(DnsRecordType recordType, CancellationToken ct)
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs]: GetGoogleDnsResult => GetGoogleDnsResultAsync

File: src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs [hunk 7]
```diff
@@ -169,62 +169,180 @@ public static async Task UpdateDnsRecordsTask(UpdateDnsRecordParameters paramete
...
+        {
+            using (var client = new RavenHttpClient { BaseAddress = new Uri(GoogleDnsApi) })
+            {
+                var response = await client.GetAsync($"/resolve?name={hostname}&type={recordType}", ct);
+                var responseString = await response.Content.ReadAsStringWithZstdSupportAsync(ct).ConfigureAwait(false);
+                
                 if (response.IsSuccessStatusCode == false)
                     throw new InvalidOperationException($"Tried to resolve '{hostname}' using Google's api ({GoogleDnsApi}).{Environment.NewLine}" +
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs]: api => API

File: src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs [hunk 8]
```diff
@@ -169,62 +169,180 @@ public static async Task UpdateDnsRecordsTask(UpdateDnsRecordParameters paramete
...
+                                                        $"Request failed with status {response.StatusCode}.{Environment.NewLine}{responseString}");
+                
                 dynamic dnsResult = JsonConvert.DeserializeObject(responseString);
-
+                
                 // DNS response format: https://developers.google.com/speed/public-dns/docs/dns-over-https
-
+                
                 if (dnsResult?.Status != 0)
                     throw new InvalidOperationException($"Tried to resolve '{hostname}' using Google's api ({GoogleDnsApi}).{Environment.NewLine}" +
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs]: api => API

File: src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs [hunk 9]
```diff
@@ -169,62 +169,180 @@ public static async Task UpdateDnsRecordsTask(UpdateDnsRecordParameters paramete
...
+        {
             try
             {
-                actualIps = (await Dns.GetHostAddressesAsync(hostname, AddressFamily.InterNetwork, cts.Token)).Select(address => address.ToString()).ToHashSet();
+                var result = (await Dns.GetHostAddressesAsync(hostname, addressFamily, ct)).Select(address => address.ToString()).ToHashSet();
+                return result;
             }
             catch (Exception e)
             {
                 throw new InvalidOperationException($"Cannot resolve '{hostname}' locally but succeeded resolving the address using Google's api ({GoogleDnsApi})." +
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs]: api => API

**Review by ppekrol — APPROVED** [review]

**Pull request merged**